src/state/slices/opportunitiesSlice/resolvers/yearn/index.ts

```diff
@@ -0,0 +1,162 @@
+import type { ToAssetIdArgs } from '@shapeshiftoss/caip'
+import { fromAssetId, toAssetId } from '@shapeshiftoss/caip'
+import { bnOrZero } from '@shapeshiftoss/investor-foxy'
+import { USDC_PRECISION } from 'constants/constants'
+import { DefiProvider, DefiType } from 'features/defi/contexts/DefiManagerProvider/DefiCommon'
+import { getYearnInvestor } from 'features/defi/contexts/YearnProvider/yearnInvestorSingleton'
+import { selectAssetById, selectPortfolioCryptoBalanceByFilter } from 'state/slices/selectors'
+
+import type {
+  GetOpportunityIdsOutput,
+  GetOpportunityMetadataOutput,
+  GetOpportunityUserStakingDataOutput,
+  OpportunitiesState,
+} from '../../types'
+import { serializeUserStakingId, toOpportunityId } from '../../utils'
+import type {
+  OpportunitiesMetadataResolverInput,
+  OpportunitiesUserDataResolverInput,
+} from '../types'
+
+export const yearnStakingOpportunitiesMetadataResolver = async ({
+  opportunityType,
+  reduxApi,
+}: OpportunitiesMetadataResolverInput): Promise<{ data: GetOpportunityMetadataOutput }> => {
+  const opportunities = await (async () => {
+    const maybeOpportunities = await getYearnInvestor().findAll()
+    if (maybeOpportunities.length) return maybeOpportunities
+
+    await getYearnInvestor().initialize()
+    return await getYearnInvestor().findAll()
+  })()
+
+  const { getState } = reduxApi
+  const state: any = getState() // ReduxState causes circular dependency
+
+  const stakingOpportunitiesById: OpportunitiesState[DefiType.Staking]['byId'] = {}
+
+  for (const opportunity of opportunities) {
+    const toAssetIdParts: ToAssetIdArgs = {
+      assetNamespace: 'erc20',
+      assetReference: opportunity.id,
+      chainId: fromAssetId(opportunity.feeAsset.assetId).chainId,
+    }
+    const assetId = toAssetId(toAssetIdParts)
+    const opportunityId = toOpportunityId(toAssetIdParts)
+
+    const asset = selectAssetById(state, assetId)
+    const underlyingAsset = selectAssetById(state, opportunity.underlyingAsset.assetId)
+
+    // Asset doesn't exist in portfolio, meaning this asset is bogus, e.g these two
+    // https://etherscan.io/address/0xa0154a44c1c45bd007743fa622fd0da4f6d67d57
+    // https://etherscan.io/address/0x5f45a578491a23ac5aee218e2d405347a0fafa8e
```

Comment:
Not applicable to Yearn, we're not expecting bogus assets there 👀  although l.53 can be kept for sanity